File: sdk/typescript/rooch-sdk/src/keypairs/ed25519/keypair.test.ts
// Copyright (c) RoochNetwork
// SPDX-License-Identifier: Apache-2.0

import { describe, it, expect } from 'vitest'

import { decodeRoochSercetKey, ROOCH_SECRET_KEY_PREFIX } from '@/crypto'

import { Ed25519Keypair } from './keypair'

const TEST_CASES: [{ pk: string; sk: string }] = [
  {
    sk: 'roochsecretkey1qqzztph49dkdl3vyw5t6fecvtuesrt9f5f2lw8ndpvqael6rr42mwulf8v7',
    pk: '3z8zMjDk70frh7I0ZF269ROuM5PeMKsYxwgFgTZEH9s=',
  },
]

describe('Ed25519 keypair', () => {
  it('create ed25519 keypair', () => {
    const kp = Ed25519Keypair.generate()

    expect(kp.getPublicKey().toBytes()).toHaveLength(32)
  })

  it('export ed25519 keypair', () => {
    const kp = Ed25519Keypair.generate()
    const secret = kp.getSecretKey()

    expect(secret.startsWith(ROOCH_SECRET_KEY_PREFIX)).toBeTruthy()
  })

  it('Create ed25519 keypair from secret key', () => {
    // valid secret key is provided by rooch keystore
    const { sk, pk } = TEST_CASES[0]

    const key = decodeRoochSercetKey(sk)
    const keypair = Ed25519Keypair.fromSecretKey(key.secretKey)

    expect(keypair.getPublicKey().toBase64()).toEqual(pk)

    const keypair1 = Ed25519Keypair.fromSecretKey(sk)

    expect(keypair1.getPublicKey().toBase64()).toEqual(pk)
  })

  it('Invalid mnemonics to derive ed25519 keypair', () => {
    expect(() => {
      Ed25519Keypair.deriveKeypair('rooch')
    }).toThrow('Invalid mnemonic')
  })

  it('Recover ed25519 keypair by mnemonics', () => {
    // mnemonics is provided by rooch cli
    // TODO:
  })

  it('Sign data', async () => {
    const keypair = new Ed25519Keypair()
    const message = new TextEncoder().encode('hello rooch')
    const signature = await keypair.sign(message)
    const isValid = keypair.getPublicKey().verify(message, signature)
    expect(isValid).toBeTruthy()
  })

  it('Sign data same as rooch cli', async () => {
    // TODO:
  })
})

File: sdk/golang/rooch-sdk/keypairs/ed25519/keypair_test.go
package ed25519

import (
	"encoding/base64"
	"testing"

	"github.com/stretchr/testify/assert"
)

var testCases = []struct {
	sk string
	pk string
}{
	{
		sk: "roochsecretkey1qqzztph49dkdl3vyw5t6fecvtuesrt9f5f2lw8ndpvqael6rr42mwulf8v7",
		pk: "3z8zMjDk70frh7I0ZF269ROuM5PeMKsYxwgFgTZEH9s=",
	},
}

func TestEd25519Keypair(t *testing.T) {
	t.Run("create ed25519 keypair", func(t *testing.T) {
		kp := GenerateEd25519Keypair()
		assert.Equal(t, 32, len(kp.GetPublicKey().ToBytes()))
	})

	t.Run("export ed25519 keypair", func(t *testing.T) {
		kp := GenerateEd25519Keypair()
		secret := kp.GetSecretKey()
		assert.True(t, strings.HasPrefix(secret, ROOCH_SECRET_KEY_PREFIX))
	})

	t.Run("Create ed25519 keypair from secret key", func(t *testing.T) {
		// valid secret key is provided by rooch keystore
		testCase := testCases[0]
		
		key, err := DecodeRoochSecretKey(testCase.sk)
		assert.NoError(t, err)
		
		keypair := Ed25519KeypairFromSecretKey(key.SecretKey)
		assert.Equal(t, testCase.pk, keypair.GetPublicKey().ToBase64())

		keypair1 := Ed25519KeypairFromSecretKey(testCase.sk)
		assert.Equal(t, testCase.pk, keypair1.GetPublicKey().ToBase64())
	})

	t.Run("Invalid mnemonics to derive ed25519 keypair", func(t *testing.T) {
		assert.Panics(t, func() {
			DeriveEd25519Keypair("rooch")
		}, "Invalid mnemonic")
	})

	t.Run("Recover ed25519 keypair by mnemonics", func(t *testing.T) {
		// TODO: Implement when mnemonics functionality is ready
		t.Skip("Not implemented yet")
	})

	t.Run("Sign data", func(t *testing.T) {
		keypair := NewEd25519Keypair()
		message := []byte("hello rooch")
		signature, err := keypair.Sign(message)
		assert.NoError(t, err)
		
		isValid := keypair.GetPublicKey().Verify(message, signature)
		assert.True(t, isValid)
	})

	t.Run("Sign data same as rooch cli", func(t *testing.T) {
		// TODO: Implement CLI signature verification
		t.Skip("Not implemented yet")
	})
} 